src/main/java/org/opensearch/security/support/StreamableRegistry.java

```diff
@@ -0,0 +1,123 @@
+import org.opensearch.common.io.stream.Writeable;
+import org.opensearch.core.common.io.stream.BaseWriteable;
+
+/**
+ * Registry for any class that does NOT implement the <code>Writeable</code> interface
+ * and needs to be serialized over the wire. Supports registration of writer and reader via <code>registerStreamable</code>
+ * for such classes and provides methods <code>writeTo</code> and <code>readFrom</code> for objects of such registered classes.
+ * <br/>
+ * Methods are protected and intended to be accessed from only within the package. (mostly by <code>Base64Helper</code>)
+ */
+public class StreamableRegistry {
+
+    private static final StreamableRegistry instance = new StreamableRegistry();
+    public final BiMap<Class<?>, Integer> classToIdMap = HashBiMap.create();
+    private final Map<Integer, Entry> idToEntryMap = new HashMap<>();
+
+    private StreamableRegistry() {
+        registerAllStreamables();
+    }
+
+    private static class Entry {
+        BaseWriteable.Writer<StreamOutput, Object> writer;
+        BaseWriteable.Reader<StreamInput, Object> reader;
+
+        Entry(BaseWriteable.Writer<StreamOutput, Object> writer, BaseWriteable.Reader<StreamInput, Object> reader) {
+            this.writer = writer;
+            this.reader = reader;
+        }
+    }
+
+    private BaseWriteable.Writer<StreamOutput, Object> getWriter(Class<?> clazz) {
+        if ( !classToIdMap.containsKey(clazz) ) {
+            throw new OpenSearchException(String.format("No writer registered for class %s", clazz.getName()));
+        }
+        return idToEntryMap.get(classToIdMap.get(clazz)).writer;
+    }
+
+    private BaseWriteable.Reader<StreamInput, Object> getReader(int id) {
+        if ( !idToEntryMap.containsKey(id) ) {
+            throw new OpenSearchException(String.format("No reader registered for id %s", id));
+        }
+        return idToEntryMap.get(id).reader;
+    }
+
+    private int getId(Class<?> clazz) {
+        if ( !classToIdMap.containsKey(clazz) ) {
+            throw new OpenSearchException(String.format("No writer registered for class %s", clazz.getName()));
+        }
+        return classToIdMap.get(clazz);
+    }
+
+    protected boolean isStreamable(Class<?> clazz) {
+        return classToIdMap.containsKey(clazz);
+    }
+
+    protected void writeTo(StreamOutput out, Object object) throws IOException {
+        out.writeByte((byte) getId(object.getClass()));
+        getWriter(object.getClass()).write(out, object);
+    }
+
+    protected Object readFrom(StreamInput in) throws IOException {
+        int id = in.readByte();
+        return getReader(id).read(in);
+    }
+
+    protected static StreamableRegistry getInstance() {
+        return instance;
+    }
+
+    protected void registerStreamable(Class<?> clazz, BaseWriteable.Writer<StreamOutput, Object> writer, BaseWriteable.Reader<StreamInput, Object> reader) {
+        Integer id = classToIdMap.size() + 1;
```

Comment:
Do these values need to be consistent over the lifetime of the cluster?  If the registration process happens in a different order on a different node what will happen?

Assuming the answers is "yes" and "bad things" we need to alter the design of this registry to handle against concurrency issues and forward combability changes.

Reply:
Yes, this needs to be consistent.

The design is similar to that of other serialization libraries such as [FST](https://github.com/RuedigerMoeller/fast-serialization) and [Kryo](https://github.com/EsotericSoftware/kryo). They too assign an incrementing ID in the order of class registration.

I've added comments for how future changes should be made to `StreamableRegistry`. Class registration methods are private to ensure that all registrations happen in one place for easy maintainability.  IMO this should suffice, but if it still is a concern, we can have statically assigned type IDs.

Reply:
Lets change the convention for this id to be based of the name of the class that is registered, this will ensure duplicates shouldn't happen based on execution order and they are easily detected if they do happen.

This also means the introduction of new types in the future won't break the ordering, does that work for you?

Reply:
This will add significant performance overhead during serialization. To keep class identification overhead at minimal, we use just a single byte to identify the class from a serialized stream.

Reply:
In fact, replacing class names with such IDs is the major contributing factor of performance in any fast serialization library.

Reply:
Thanks for discussion this change with me, this is the single place for all registration so the concerns about consistency aren't very worrisome since developers would be modifying this class and be aware of its implications on registered Streamable